### PR TITLE
Add internal docs for staticdata.c

### PR DESCRIPTION
**Pull request by timholy**

### CODE REVIEW

**Review by jpsamaroo — COMMENTED** [review]
Awesome!

File: src/staticdata.c [hunk 1]
```diff
@@ -2,6 +2,44 @@
...
+    so recursion is mimicked using a work-queue managed by `jl_serialize_reachable`.
+
+    It's worth emphasizing that despite the name `jl_serialize_value`, the only goal of this stage is to
+    insert objects into `backref_table`, specifically the objects/fields that will require touch-up when the system
+    image is loaded.
+
+  - step 2 (the biggest of four steps) takes all items in `backref_table` and actually serializes them.
+    They are first ordered by `id` and then serialized. Serialization constructs `layout_table`, where `layout_table[id]` is
+    the stream position of the item corresponding to `id`. There are actually several distinct streams (see the
+    fields in `jl_serializer_state`) that reflect the format of ELF/LLVM; for example, global variables get added to the
```
[the remark this answers is not in the document]
Comment (timholy) [reply src/staticdata.c:26]: Would "conceptual organization" be clearer, or should I just drop LLVM?

File: src/staticdata.c [hunk 2]
```diff
@@ -51,15 +89,7 @@ jl_value_t **const*const get_tags(void) {
...
-        INSERT_TAG(jl_tvar_type);
-        INSERT_TAG(jl_method_instance_type);
-        INSERT_TAG(jl_method_type);
-        INSERT_TAG(jl_code_instance_type);
-        INSERT_TAG(jl_linenumbernode_type);
-        INSERT_TAG(jl_lineinfonode_type);
-        INSERT_TAG(jl_gotonode_type);
-        INSERT_TAG(jl_quotenode_type);
-        INSERT_TAG(jl_gotoifnot_type);
+        INSERT_TAG(jl_tvar_type);reinit
```
Comment (vtjnash) [inline src/staticdata.c:92]:
```suggestion
        INSERT_TAG(jl_tvar_type);
        INSERT_TAG(jl_method_instance_type);
        INSERT_TAG(jl_method_type);
        INSERT_TAG(jl_code_instance_type);
        INSERT_TAG(jl_linenumbernode_type);
        INSERT_TAG(jl_lineinfonode_type);
        INSERT_TAG(jl_gotonode_type);
        INSERT_TAG(jl_quotenode_type);
        INSERT_TAG(jl_gotoifnot_type);
```

File: src/staticdata.c [hunk 3]
```diff
@@ -218,15 +248,13 @@ static uintptr_t nsym_tag;
...
-// table of all objects that are serialized
-static htable_t backref_table;
-static int backref_table_numel;
-static arraylist_t layout_table;
-static arraylist_t object_worklist;
-
-// list of (size_t pos, (void *f)(jl_value_t*)) entries
-// for the serializer to mark values in need of rework by function f
-// during deserialization later
+// table of all objects that are serializeds->relocs
```
Comment (vtjnash) [inline src/staticdata.c:251]:
```suggestion
// table of all objects that are serialized
static htable_t backref_table;
static int backref_table_numel;
static arraylist_t layout_table;
static arraylist_t object_worklist;

```

**Review by vtjnash — COMMENTED** [review]
SGTM

File: src/staticdata.c [hunk 4]
```diff
@@ -2,6 +2,44 @@
 
 /*
   saving and restoring system images
+
+  This performs serialization and deserialization, somewhat similar to dump.c. However, while dump.c is exhaustive,
+  here we focus on issues like pointer relocation and registering with the garbage collector. We also need to pay special
+  attention to things like builtin functions, C-implemented types (those in jltypes.c), the metadata for
+  documentation,
```
Comment (vtjnash) [inline src/staticdata.c:9]:
```suggestion
  This performs serialization and deserialization of in-memory data. The dump.c file is similar, but has less complete coverage than this. One main difference is that this format focuses on deserialization performance, while dump.c is focused on being a single pass, but suffers therefore from extra complexity. On deserialization, we only need to deal with pointer relocation, registering with the garbage collector, and making note of special internal types. During serialization, we also need to pay special attention to things like builtin functions, C-implemented types (those in jltypes.c), the metadata for documentation, optimal layouts, integration with native system image generation, and preparing other preprocessing directives.
```

File: src/staticdata.c (hunk 2 above)
Comment (timholy) [reply src/staticdata.c:92]: Ah, the hazards of using the online VSCode editor...

**Pull request merged**